buildSrc - move group/version logic to org.jetbrains.androidx.build

igordmn opened the pull request:
- restore group/version logic to AOSP state
- move the JetBrains libraries declaration to `buildSrc/.../JetBrainsPublication.kt` (it serves as a data for all modules, similar to `libraryversions.toml`)
- add logic to replace group/version after `AndroidXPlugin` is applied

## Testing
1. CI Snapshot build is [green](https://teamcity.jetbrains.com/buildConfiguration/JetBrainsPublicProjects_Compose_AllPersonalBuild/5549169)
2. Compare `./m2` files after
```
./gradlew publishComposeJbToMavenLocal -Pcompose.platforms=all
```
3. Compare `./m2` files after 
```
./gradlew publishComposeJbToMavenLocal -Pcompose.platforms=all -Pjetbrains.publication.libraries=COMPOSE,COMPOSE_MATERIAL_NAVIGATION,COMPOSE_MATERIAL3,COMPOSE_MATERIAL3_ADAPTIVE,LIFECYCLE,NAVIGATION_3,NAVIGATION_EVENT,SAVEDSTATE,WINDOW -Pjetbrains.publication.libraries=COMPOSE,COMPOSE_MATERIAL_NAVIGATION,COMPOSE_MATERIAL3,COMPOSE_MATERIAL3_ADAPTIVE,LIFECYCLE,NAVIGATION_3,NAVIGATION_EVENT,SAVEDSTATE,WINDOW -Pjetbrains.publication.version.COMPOSE=1.10.0+dev546 -Pjetbrains.publication.version.COMPOSE_MATERIAL_NAVIGATION=1.10.0+dev546 -Pjetbrains.publication.version.COMPOSE_MATERIAL3=1.10.0+dev546 -Pjetbrains.publication.version.COMPOSE_MATERIAL3_ADAPTIVE=1.2.0+dev546 -Pjetbrains.publication.version.LIFECYCLE=2.10.0+dev546 -Pjetbrains.publication.version.NAVIGATION_3=1.0.0+dev546 -Pjetbrains.publication.version.NAVIGATION_EVENT=1.0.0+dev546 -Pjetbrains.publication.version.SAVEDSTATE=1.4.0+dev546 -Pjetbrains.publication.version.WINDOW=1.5.0+dev546
```
4. Sync IDE
5. Run run1, Demo (desktop/web)

## Release Notes
N/A

igordmn committed:
Reset code to AOSP state

igordmn committed:
Disable non-used publication functionality

Otherwise, it fails

igordmn committed:
Change group and version

igordmn commented on buildSrc/public/src/main/kotlin/androidx/build/Version.kt:
```diff
@@ -77,7 +77,7 @@ data class Version(
 
         private val VERSION_FILE_REGEX = Pattern.compile("^(res-)?(.*).txt$")
-        private val VERSION_REGEX = Pattern.compile("^(\\d+)\\.(\\d+)\\.(\\d+)((?:\\.\\d+)?(?:[-\\+].+)?)?$")
+        private val VERSION_REGEX = Pattern.compile("^(\\d+)\\.(\\d+)\\.(\\d+)(-.+)?$")
```
It was changed because it doesn't support `1.9.0+dev534` format. But now we don't use this class for versioning.

It might be needed if we want to reuse some checks (API check, dependencies check), but we can look at it when we introduce them.

igordmn replied:
Reverted for now, it is needed

igordmn commented on buildSrc/public/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsPublication.kt:
```diff
@@ -0,0 +1,184 @@
+ * building the JetBrains fork of AOSP.
+ */
+object JetBrainsPublication {
+    val libraryToComponents = mapOf(
```
Moved from the `mpp` module with these changes:
- removed TODO
- removed the commented material-icons
- added ":compose:desktop:desktop"

igordmn commented on compose/runtime/runtime-tracing/build.gradle:
```diff
@@ -45,7 +45,6 @@ dependencies {
 androidx {
     name = "Compose Runtime: Tracing"
     publish = Publish.SNAPSHOT_AND_RELEASE
-    mavenVersion = LibraryVersions.COMPOSE_RUNTIME_TRACING
```
Reset this file to the AOSP state

igordmn commented on buildSrc/private/src/main/kotlin/androidx/build/AndroidXExtension.kt:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2023 The Android Open Source Project
+ * Copyright 2018 The Android Open Source Project
  *
```
Reset this file to the AOSP state

igordmn commented on buildSrc/private/src/main/kotlin/androidx/build/LibraryVersionsService.kt:
```diff
@@ -36,7 +36,6 @@ abstract class LibraryVersionsService : BuildService<LibraryVersionsService.Para
         var composeCustomVersion: Provider<String>
```
Reset this file to the AOSP state

igordmn commented on buildSrc/private/src/main/kotlin/org/jetbrains/androidx/build/MavenUploadHelper.kt:
```diff
@@ -18,16 +18,13 @@ package org.jetbrains.androidx.build
 
 import androidx.build.AndroidXExtension
```
Remove code from this file that isn't used in our publication. It depends on the `androidx` versions, not the overriden project versions

igordmn commented on libraryversions.toml:
```diff
@@ -97,8 +97,8 @@ MEDIA = "1.7.0-rc01"
 MEDIAROUTER = "1.8.0-alpha01"
 METRICS = "1.0.0-beta02"
```
Reset this file to the AOSP state

MatkovIvan commented on libraryversions.toml:
```diff
@@ -97,8 +97,8 @@ MEDIA = "1.7.0-rc01"
-NAVIGATION_3 = "1.0.0-alpha10"
-NAVIGATION_EVENT = "1.0.0-alpha09"
+NAVIGATION3 = "1.0.0-alpha10"
+NAVIGATIONEVENT  = "1.0.0-alpha09"
```
Double space
```suggestion
NAVIGATIONEVENT = "1.0.0-alpha09"
```

MatkovIvan commented on libraryversions.toml:
```diff
@@ -311,55 +310,3 @@ WINDOW_EXTENSIONS_CORE = { group = "androidx.window.extensions.core", atomicGrou
-COLLECTION_JETBRAINS = { group = "org.jetbrains.compose.collection-internal", atomicGroupVersion = "versions.COMPOSE", overrideInclude = [ ":collection:collection" ] }
-COMPOSE_COMPILER = { group = "org.jetbrains.compose.compiler", atomicGroupVersion = "versions.COMPOSE" }
-COMPOSE_DESKTOP = { group = "org.jetbrains.compose.desktop", atomicGroupVersion = "versions.COMPOSE" }
-COMPOSE_MATERIAL_NAVIGATION = { group = "org.jetbrains.compose.material", atomicGroupVersion = "versions.COMPOSE_MATERIAL_NAVIGATION", overrideInclude = [ ":compose:material:material-navigation" ] }
```
It seems `versions.COMPOSE_MATERIAL_NAVIGATION` is unused now and should be removed

igordmn replied:
Done

MatkovIvan replied:
[CMP-8861](https://youtrack.jetbrains.com/issue/CMP-8861) fixed now?

igordmn replied:
No, we should change it in the JetBrainsPublication (new class). I will do it later in another PR

MatkovIvan commented on mpp/build.gradle.kts:
```diff
@@ -1,7 +1,7 @@
+import org.jetbrains.androidx.build.JetBrainsPublication
 import org.jetbrains.androidx.build.artifactRedirection
 import org.jetbrains.androidx.build.hasRedirection
 
```
Do we still need `buildscript` below this line here? It triggers some custom error checks with new buildSrc and removing it, seem works just fine

igordmn replied:
It is not needed. Removed.

igordmn committed:
Remove Androidx backhandler publication

igordmn committed:
Remove androidx.alternativeProjectUrl

igordmn committed:
Remove buildscript in mpp

igordmn committed:
Remove COMPOSE_MATERIAL_NAVIGATION

igordmn committed:
Update libraryversions.toml

Co-authored-by: Ivan Matkov <ivan.matkov@jetbrains.com>

igordmn committed:
Restore backhandler publication and version parsing

SNAPSHOT_AND_RELEASE adds release Compose compiler arguments:
```
plugin:androidx.compose.compiler.plugins.kotlin:sourceInformation=true
```

Also, another code cast the version to Version when it is SNAPSHOT_AND_RELEASE

MatkovIvan commented on buildSrc/public/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsPublication.kt:
```diff
@@ -0,0 +1,186 @@
+    }
+}
+
+fun ComposeComponent.library() = projectPathToLibrary[path]!!
```
Maybe it's worth to use `requireNotNull` with readable message here

igordmn replied:
Done

MatkovIvan commented on buildSrc/public/src/main/kotlin/org/jetbrains/androidx/build/JetBrainsPublication.kt:
```diff
@@ -0,0 +1,186 @@
+    }
+
+    companion object {
+        fun isValid(libraryName: String) =
```
Sounds too generic, it's not clear without deep context what's exacty it checks

igordmn replied:
Done

igordmn committed:
Feedback
